### PR TITLE
fix healing ability of some characters to some extent

### PR DESCRIPTION
**Pull request by saucebing**
## Description

fix healing ability of some characters to some extent

## Type of changes

- [x] Bug fix
- [ ] New feature 
- [ ] Enhancement
- [ ] Documentation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My pull request is unique and no other pull requests have been opened for these changes
- [x] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [x] I am responsible for any copyright issues with my code if it occurs in the future.

### CODE REVIEW

**Comment by zhaodice** [comment]
see #1202

File: src/main/java/emu/grasscutter/data/ResourceLoader.java [hunk 1]
```diff
@@ -269,7 +269,15 @@ private static void loadAbilityModifiers() {
 			List<AbilityConfigData> abilityConfigList;
 			
 			try (FileReader fileReader = new FileReader(file)) {
-				abilityConfigList = Grasscutter.getGsonFactory().fromJson(fileReader, TypeToken.getParameterized(Collection.class, AbilityConfigData.class).getType());
+                BufferedReader bufferedReader = new BufferedReader(fileReader);
+                String line = "", lines = "";
+                while((line = bufferedReader.readLine()) != null){
+                    lines += line;
+                }
+                lines = lines.replaceAll("HNEIEGHMLKH", "modifiers");
```
Comment (Akka0) [inline src/main/java/emu/grasscutter/data/ResourceLoader.java]: It's recommended to use like `@SerializedName(value="x", alternate={"_x", "X"})` in data class instead, see `Position` class

File: src/main/java/emu/grasscutter/game/ability/AbilityManager.java [hunk 2]
```diff
@@ -152,22 +160,65 @@ private void handleGenerateElemBall(AbilityInvokeEntry invoke) throws InvalidPro
...
+    }
+
+	private void invokeAction(AbilityModifierAction action, GameEntity target, GameEntity sourceEntity, String modifierString) {
 		switch (action.type) {
 			case HealHP -> {
-				if (action.amount == null) {
-					return;
-				}
-				
+                ArrayList<HealData> healDataList = new ArrayList();
```
Comment (ghost) [inline src/main/java/emu/grasscutter/game/ability/AbilityManager.java]: While this is a step in the right direction, variables from HealData are dynamic and recieved from the client so there is no need to hard code every character's healing amount.

**Review by ghost — APPROVED** [review]

**Review by ghost — CHANGES_REQUESTED** [review]
Didn't mean to approve, my mistake. Please resolve mine and @Akka0's requested changes first.

**Comment by saucebing** [comment]
I have resolved @Akka0 's requested changes. As for @Melledy 's, maybe I need to sniff. However, it need some time. I will be appreciated if anyone can help me or provide some guidance.

**Comment by ghost** [comment]
> I have resolved @Akka0 's requested changes. As for @Melledy 's, maybe I need to sniff. However, it need some time. I will be appreciated if anyone can help me or provide some guidance.

The client sends ability variables and their values to the server, which are handled and stored in `entity::getMetaOverrideMap()`

**Comment by saucebing** [comment]
> > I have resolved @Akka0 's requested changes. As for @Melledy 's, maybe I need to sniff. However, it need some time. I will be appreciated if anyone can help me or provide some guidance.
> 
> The client sends ability variables and their values to the server, which are handled and stored in `entity::getMetaOverrideMap()`

Thank you for your guidances, Melledy. In invokeAction function, I tried to check keys and values in entity::getMetaOverrideMap(), and found that it did store the healing base and percentage of some skills, such as E skills of Kokomi and Qiqi. However, it still lacks of some other skills, such as Q skill of Qiqi, in which there is nothing in entity::getMetaOverrideMap(). Do you know the possible reason? Or do you have any suggestion?

**Comment by zhaodice** [comment]
> Thank you for your guidances, Melledy. In invokeAction function, I tried to check keys and values in entity::getMetaOverrideMap(), and found that it did store the healing base and percentage of some skills, such as E skills of Kokomi and Qiqi. However, it still lacks of some other skills, such as Q skill of Qiqi, in which there is nothing in entity::getMetaOverrideMap(). Do you know the possible reason? Or do you have any suggestion?

Could you supply information about Q of Qiqi to getMetaOverrideMap?

**Comment by saucebing** [comment]
> > Thank you for your guidances, Melledy. In invokeAction function, I tried to check keys and values in entity::getMetaOverrideMap(), and found that it did store the healing base and percentage of some skills, such as E skills of Kokomi and Qiqi. However, it still lacks of some other skills, such as Q skill of Qiqi, in which there is nothing in entity::getMetaOverrideMap(). Do you know the possible reason? Or do you have any suggestion?
> 
> Could you supply information about Q of Qiqi to getMetaOverrideMap?

I am trying, or I will use some other methods to try to solve it.

**Comment by saucebing** [comment]
I have resolved @Melledy 's requested changes.
Now I have fixed:
(1) E skills of Kokomi, Noel, Barbara, Shinobu and Qiqi; 
(2) Q skills of Qin, Bernett, Diona and Sayu.

**Review by ghost — APPROVED** [review]

**Comment by Cardgodk** [comment]
> 我已解决 @Melledy 请求的更改。现在我已经修复了：（1）Kokomi，Noel，Barbara，Shinobu和Qiqi的E技能; （2）秦、伯内特、迪奥娜、萨尤的Q技能。

Hello, I found a small problem. Kokomi's e skill recovery value should benefit from Kokomi's own HP, not the maximum HP of the character she treats

**Comment by Cardgodk** [comment]
> 

Because I haven't done a lot of testing, I'm not sure if Kokomi is the only one with this problem

**Comment by saucebing** [comment]
> > 我已解决 @Melledy 请求的更改。现在我已经修复了：（1）Kokomi，Noel，Barbara，Shinobu和Qiqi的E技能; （2）秦、伯内特、迪奥娜、萨尤的Q技能。
> 
> Hello, I found a small problem. Kokomi's e skill recovery value should benefit from Kokomi's own HP, not the maximum HP of the character she treats

I have found this problem, which will be solved in the next pr.

**Comment by saucebing** [comment]
> Kokomi's e skill recovery value should benefit from Kokomi's own HP, not the maximum HP of the character she treats

It is fixed in the latest pr.

**Pull request merged**